### PR TITLE
Add a public accessor for the props a request selected

### PR DESCRIPTION
**Pull request by qarlosh**
New public API: a way to read what a request selected, from a view that has to build its queryset before the filtering runs.

## Why

Django Rest Framework calls `get_queryset()` before any filter backend, so a view that decides which prefetches or annotations are worth adding based on `select(...)` has nothing to read at that point. The parsed selects were reachable in exactly one place, `request.rql_select`, which the backend only sets while filtering and only for filter classes with `SELECT = True`. Turning `SELECT` on is not always an option either: it makes the filter class validate select names and reject any it does not declare, which is a breaking change for an established collection whose clients already send names of their own.

The result was consumers re-deriving from the query string something the library already knows how to read.

## What this adds

```py3
from dj_rql.drf import get_select_props


class ProductViewSet(mixins.ListModelMixin, GenericViewSet):
    filter_backends = (RQLFilterBackend,)
    rql_filter_class = ProductFilters

    def get_queryset(self):
        queryset = Product.objects.all()

        if 'reviews' in get_select_props(self.request):
            queryset = queryset.prefetch_related('reviews')

        return queryset
```

- **`dj_rql.transformer.RQLSelectTransformer`**, next to `RQLLimitOffsetTransformer`: collects the props of an AST and nothing else.
- **`dj_rql.drf.get_select_props(request)`**: reuses the AST filtering left on the request when there is one and parses the query otherwise, the way `RQLLimitOffsetPagination` already does. Parsing twice is cheap: the parser keeps a process-wide cache, so the second call is a cache hit.

The query is the one the filtering will be given, not the one that was sent: it is resolved through the RQL backend of the view, which matters for any backend that rewrites it, such as the compatibility ones. Django Rest Framework leaves the view on the request, so the caller passes nothing.

Deliberate properties, all covered by tests: props come back as the query wrote them, with `-` prefixes, dotted paths whole, order and repetitions preserved and nothing deduplicated; nothing is validated against a filter class, so it works with `SELECT = False` and never turns an unknown name into a rejection; and a query that cannot be parsed yields no props instead of raising, so the accessor is never the reason a request fails before the filtering has had its say. Rejecting a bad query stays entirely the job of `apply_filters`.

## Behaviour change worth a release note

A query holding two `select(...)` operations was rejected through a bare `assert`, so it surfaced as the generic `"Bad filter query."`. It now says:

> `Bad select filter: query can contain only one select operation.`

matching what `ordering` has always reported for the same restriction. Anything asserting the old string will need updating.

Two queries that used to be accepted are now rejected, so this is a behaviour change and not only a wording one. `select()&select(author)` and `select(author)&select()` are the same query, yet only the second was rejected: the check looked at the truthiness of what the first select collected, and an empty one collected nothing. Both are rejected now, under plain CPython. And the restriction was a bare `assert`, which `python -O` strips, so under optimised bytecode any pair of select operations was silently accepted, the last one winning; those are rejected now too. Merging multiple select operations was considered and deliberately not done: one operation per query is the library's convention, and `ordering` enforces it the same way.

## Notes

- Public API on a library, so this warrants a minor version. The version in `pyproject.toml` is left alone, since the release tag sets it.
- Documented in `docs/select.md` (a section on reading what a request selected) and in `docs/reference.md`.
- 548 tests pass, linter clean, and both new files are fully covered. Independent of #120 and #121: no file in common, so it can be reviewed and merged in any order.

### CODE REVIEW

File: dj_rql/drf/selects.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        hold only one; and a query that cannot be parsed returns no props at all,
+        instead of raising. Rejecting either of those stays the job of
+        `RQLFilterClass.apply_filters`, which runs later, so that this never becomes the
+        reason a request fails before it.
+    """
+    try:
+        rql_ast = getattr(drf_request, 'rql_ast', None)
+
+        if rql_ast is None:
+            query = get_query(drf_request)
```
Comment (pcaro) [inline dj_rql/drf/selects.py:42]: [SOL-A-UNX] `get_select_props` reads the raw `QUERY_STRING`, bypassing the backend's overridable `get_query`. With `DjangoFiltersRQLFilterBackend`, `?title__in=a,b&select(author)` returns `()` in `get_queryset()` but `('author',)` after filtering, so the view skips the Prefetch and we get the N+1 this helper exists to avoid. Should it resolve the query through the view's backend instead?

File: dj_rql/drf/selects.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+            if not query:
+                return ()
+
+            rql_ast = RQLParser.parse_query(query)
+
+        return RQLSelectTransformer().transform(rql_ast)
+    except (LarkError, RQLFilterError):
```
Comment (pcaro) [inline dj_rql/drf/selects.py:49]: [SOL-A-PTE] The docstring says it never raises, but only `LarkError`/`RQLFilterError` are caught. DRF's schema generators (and drf-spectacular) call `get_queryset()` with `view.request = None`, which raises `AttributeError` here. Either guard (`getattr` / `META.get('QUERY_STRING', '')`) or narrow the docs claim.

Comment (pcaro) [inline dj_rql/drf/selects.py:49]: [SOL-B-STD] Catching `LarkError` only ever matches `VisitError` from our own transformer callbacks, so a bug in `RQLSelectTransformer` becomes "no props" and a silent N+1 with a 200. `apply_filters` unwraps and re-raises these. I'd catch only `RQLFilterParsingError` around `parse_query`; the `LarkError` import and the `if not query` guard then go away.

File: dj_rql/drf/selects.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        of any filter class. It is up to the caller to decide what a prop means to it.
+
+        Props of every select operation are returned, even though a query is allowed to
+        hold only one; and a query that cannot be parsed returns no props at all,
+        instead of raising. Rejecting either of those stays the job of
+        `RQLFilterClass.apply_filters`, which runs later, so that this never becomes the
+        reason a request fails before it.
+    """
+    try:
+        rql_ast = getattr(drf_request, 'rql_ast', None)
```
Comment (pcaro) [inline dj_rql/drf/selects.py:39]: [SOL-B-REF] Third copy of "reuse `request.rql_ast`, else parse", and the fallback differs from `RQLLimitOffsetPagination` (that one treats `rql_ast=None` as "skip", this one re-parses). Can we extract a single `get_rql_ast(drf_request)` next to `get_query` and use it in the backend, pagination and here?

File: dj_rql/drf/selects.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+        if rql_ast is None:
+            query = get_query(drf_request)
+            if not query:
+                return ()
+
+            rql_ast = RQLParser.parse_query(query)
+
+        return RQLSelectTransformer().transform(rql_ast)
```
Comment (pcaro) [inline dj_rql/drf/selects.py:48]: [COD-C-CMT] Nothing is memoised, so each `'x' in get_select_props(...)` re-runs the transform. Minor, but the docs should say "call once, keep the tuple", or stash the result on the request like `rql_ast`.

File: dj_rql/transformer.py [hunk 5]
```diff
@@ -176,7 +177,12 @@ def ordering(self, args):
         return self._q()
 
     def select(self, args):
-        assert not self._select
+        if self._select:
```
Comment (pcaro) [inline dj_rql/transformer.py:180]: [COD-A-CRN] `if self._select:` checks list truthiness, so `select()&select(author)` is accepted while `select(author)&select()` is rejected. Use a `None` sentinel (`if self._select is not None`) and add the empty-first case to `test_several_select_operations`.

File: dj_rql/transformer.py [hunk 6]
```diff
@@ -176,7 +177,12 @@ def ordering(self, args):
...
+            raise RQLFilterParsingError(
```
Comment (pcaro) [inline dj_rql/transformer.py:181]: [SOL-B-STD] The PR description says "still rejected, exactly as before — only the message changes". Under `python -O` the old `assert` was stripped, so double selects were silently accepted (last one wins) and now become 400s. The release note should state this as a behaviour change.

File: docs/select.md [hunk 7]
```diff
@@ -120,3 +120,51 @@ class ProductSerializer(RQLMixin, serializers.ModelSerializer):
...
+
+
+class ProductViewSet(mixins.ListModelMixin, GenericViewSet):
+    filter_backends = (RQLFilterBackend,)
+    rql_filter_class = ProductFilters
+
+    def get_queryset(self):
+        queryset = Product.objects.all()
+
+        if 'reviews' in get_select_props(self.request):
```
Comment (pcaro) [inline docs/select.md:145]: [SOL-B-UNX] `if 'reviews' in get_select_props(self.request)` misses nested selects: `select(author.email)` yields `('author.email',)`, but dj_rql treats it as selecting `author` and the serializer renders it without the Prefetch. Document that props are raw dotted paths and show a prefix-aware check, or expose the include/exclude split `_prepare_selects` already computes.

File: tests/test_filter_cls/test_apply_filters.py [hunk 8]
```diff
@@ -23,7 +23,7 @@ def apply_filters(query):
     return list(q)
 
 
-@pytest.mark.parametrize('bad_query', ['q=', '(select(stats.attributes))&select(stats.attributes)'])
+@pytest.mark.parametrize('bad_query', ['q='])
```
Comment (pcaro) [inline tests/test_filter_cls/test_apply_filters.py:26]: [COD-C-ALT] Moving the double-select case out of `test_parsing_error` leaves `filter_cls.py:272` (generic `raise RQLFilterParsingError()`) uncovered; nothing in the suite reaches it now. Add a case that provokes a `VisitError` wrapping `AssertionError`, or drop the dead arm.

File: tests/dj_rf/view.py [hunk 9]
```diff
@@ -67,6 +68,20 @@ def get_rql_filter_class(self):
         return SelectBooksFilterClass
 
 
+select_props_seen = []
```
Comment (pcaro) [inline tests/dj_rf/view.py:71]: [COD-C-ALT] Module-level `select_props_seen = []` shared with the test module needs `.clear()` in three tests and is the only such global in the suite. `mocker.spy(tests.dj_rf.view, 'get_select_props')` + `spy.spy_return` gives the same assertion with no global (pytest-mock is already a dev dep).

**Review by pcaro — CHANGES_REQUESTED** [review]
Three blockers (`SOL-A` x2 in `selects.py`, `COD-A` in `transformer.py`); the rest can be discussed or go as follow-ups.

File: dj_rql/drf/selects.py (hunk 1 above)
Comment (qarlosh) [reply dj_rql/drf/selects.py]: Fixed, and thanks for the catch. The accessor now asks the view's RQL backend for the query, through a new `RQLFilterBackend.get_query_for_view`. DRF leaves the view on `request.parser_context`, so no signature change and no work for the caller; without a view or an RQL backend it falls back to the query string. This also covers any custom `get_query` override, not just the shipped compatibility backends. Test added behind `DjangoFiltersRQLFilterBackend` with your exact query.

File: dj_rql/drf/selects.py (hunk 2 above)
Comment (qarlosh) [reply dj_rql/drf/selects.py]: Fixed: returns `()` when the request is `None`, with a test naming the schema generators.

Comment (qarlosh) [reply dj_rql/drf/selects.py]: Agreed. Narrowed to `RQLFilterParsingError` around the parse, so a `VisitError` from our own transformer propagates instead of turning into a silent N+1. Kept the `if not query` guard though: without it every request with no query string pays a failed parse plus an exception.

File: dj_rql/drf/selects.py (hunk 3 above)
Comment (qarlosh) [reply dj_rql/drf/selects.py]: Follow-up, if you agree. The backend does not share the pattern (it always parses through `apply_filters`), so it is two copies rather than three, and unifying them changes pagination's deliberate "no AST, skip" behaviour. Rather not fold a refactor of working code into a new-API PR.

File: dj_rql/drf/selects.py (hunk 4 above)
Comment (qarlosh) [reply dj_rql/drf/selects.py]: Documented: keep the returned tuple rather than calling once per prop. Not caching it on the request, that is state and invalidation for very little.

File: dj_rql/transformer.py (hunk 5 above)
Comment (qarlosh) [reply dj_rql/transformer.py]: Fixed with a `None` sentinel, and `select_filters` normalises back to `[]` for its callers. Both orders are in the test now.

File: dj_rql/transformer.py (hunk 6 above)
Comment (qarlosh) [reply dj_rql/transformer.py:183]: Right, and I had missed it. Added to the description: under `-O` the assert was stripped, so double selects were silently accepted, last one winning. That is a behaviour change, not just a message.

File: docs/select.md (hunk 7 above)
Comment (qarlosh) [reply docs/select.md]: Fixed: the docs now say props are raw dotted paths and that a prop selects everything below it, and the example tests the prefix instead of equality. Not exposing the `_prepare_selects` split here, that is its own API decision.

File: tests/test_filter_cls/test_apply_filters.py (hunk 8 above)
Comment (qarlosh) [reply tests/test_filter_cls/test_apply_filters.py:26]: Restored, you are right that I dropped it. New test with a custom filter raising `AssertionError` inside the transform, which reaches that arm deterministically.

File: tests/dj_rf/view.py (hunk 9 above)
Comment (qarlosh) [reply tests/dj_rf/view.py]: Done, `mocker.spy` + `spy_return`. The global and the three `.clear()` calls are gone.

File: dj_rql/drf/selects.py (hunk 3 above)
Comment (pcaro) [reply dj_rql/drf/selects.py]: I Agree

File: docs/select.md [hunk 10]
```diff
@@ -120,3 +120,59 @@ class ProductSerializer(RQLMixin, serializers.ModelSerializer):
...
+## Reading what a request selected
+
+Sometimes a view has to know what the caller selected *before* the filtering runs,
+because the queryset it builds depends on it. A typical case is deciding which
+`Prefetch` objects or annotations are worth adding, which happens in `get_queryset()`,
+and Django Rest Framework calls it before the filter backend gets to run.
+
+`dj_rql.drf.get_select_props` reads the props of the request, asking the RQL backend of
+the view for the query, so it gets the same one the filtering will be given even when
+the backend rewrites it, as the [compatibility backends](#) do:
```
Comment (pcaro) [inline docs/select.md:133]: [COD-C-CMT] `as the [compatibility backends](#) do` — the anchor is a placeholder, and it's the only `](#)` in `docs/`. There's nothing in `docs/` describing the compatibility backends to point at either (no page, no heading), so the link has no target to gain. Drop the markup and leave the plain words, unless you want to link `reference.md#filter-backend-dj_rqldrfbackendrqlfilterbackend` — which documents `RQLFilterBackend`, not the compat ones, so it would be a slightly wrong target.

File: dj_rql/drf/backend.py [hunk 11]
```diff
@@ -115,6 +115,30 @@ def get_filter_class(view):
     def get_query(cls, filter_instance, request, view):
         return get_query(request)
 
+    @classmethod
```
Comment (pcaro) [inline dj_rql/drf/backend.py:118]: [SOL-C-CMT] `get_query_for_view` is a public classmethod on a public backend class, with a docstring written for someone overriding it — but `docs/reference.md:149` pins `members: [filter_queryset]`, so it renders nowhere. Either add it to that list, or make it `_get_query_for_view` if resolving the query is meant to stay the accessor's private business. Right now it reads as API and ships as neither.

File: tests/test_filter_cls/test_apply_filters.py [hunk 12]
```diff
@@ -357,6 +381,24 @@ def test_several_ordering_operations():
     assert e.value.details['error'] == expected
 
 
+@pytest.mark.parametrize(
+    'bad_query',
+    (
+        'select(author)&select(published.at)',
+        '(select(stats.attributes))&select(stats.attributes)',
+        # An empty select is still a select operation.
+        'select()&select(author)',
```
Comment (pcaro) [inline tests/test_filter_cls/test_apply_filters.py:390]: [SOL-B-STD] These two cases are the second half of the behaviour change, and the description only tells the first. `select()&select(author)` was accepted under plain CPython — no `-O` needed, the old `if self._select:` just didn't see an empty select — and is now a 400. The release note covers the `-O` path but not this one, which is the one every user is actually running. Worth a line in the same section.

File: tests/test_drf/test_select_props.py [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+
+
+@pytest.mark.django_db
+def test_props_are_available_while_building_the_queryset(api_client, clear_cache, mocker):
+    spy = mocker.spy(tests.dj_rf.view, 'get_select_props')
+
+    response = api_client.get('/select_props/?select(author,-pages)&eq(title,book)')
+
+    assert response.status_code == 200
+    assert spy.spy_return == ('author', '-pages')
```
Comment (pcaro) [inline tests/test_drf/test_select_props.py:141]: [COD-C-ALT] Nit: `spy_return` is the last call's value, so this no longer pins how many times `get_queryset()` ran; the old `select_props_seen == [('author', '-pages')]` did. An extra `spy.assert_called_once()` here gets it back. Only worth it in this first test, the others don't care.

**Review by pcaro — APPROVED** [review]
All ten threads addressed, checked against `007e050` rather than taken on trust:

- The accessor now resolves the query through the view's RQL backend, and the compatibility backend's own `RQLFilterParsingError` stays inside the caught scope, so it still cannot fail a request before the filtering does.
- Narrowing to `RQLFilterParsingError` is right: `RQLLarkParser.parse_query` already converts its own `LarkError`, so a `VisitError` from our transformer propagates as it should.
- The `None` sentinel is consistent with `select_filters` feeding `_build_select_data`, and both orders are covered.
- `filter_cls.py` reaches the generic parsing arm again; the module-level global is gone.

Locally: 554 passed, flake8 clean, `selects.py`, `backend.py` and `transformer.py` at 100%.

Approving. The four comments I left are two documentation fixes, one missing line in the release note and a test nit — none of them blocking, so take them or leave them before merging.
